test/model/compose_test.dart

```diff
@@ -0,0 +1,391 @@
+    });
+
+    test('four-backtick block', () {
+      final content = chomp('''
+````
+hello world
+````
+''');
+      final expected = chomp('''
+`````
+````
+hello world
+````
+`````
+''');
+      final expectedWithInfoString = chomp('''
+`````info
+````
+hello world
+````
+`````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('five-backtick block', () {
+      final content = chomp('''
+`````
+hello world
+`````
+''');
+      final expected = chomp('''
+``````
+`````
+hello world
+`````
+``````
+''');
+      final expectedWithInfoString = chomp('''
+``````info
+`````
+hello world
+`````
+``````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('three-, four-, and five-backtick blocks', () {
+      final content = chomp('''
+```
+hello world
+```
+
+````
+hello world
+````
+
+`````
+hello world
+`````
+''');
+      final expected = chomp('''
+``````
+```
+hello world
+```
+
+````
+hello world
+````
+
+`````
+hello world
+`````
+``````
+''');
+      final expectedWithInfoString = chomp('''
+``````info
+```
+hello world
+```
+
+````
+hello world
+````
+
+`````
+hello world
+`````
+``````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('dangling opening fence', () {
+      final content = chomp('''
+````javascript
+// hello world
+''');
+      final expected = chomp('''
+`````
+````javascript
+// hello world
+`````
+''');
+      final expectedWithInfoString = chomp('''
+`````info
+````javascript
+// hello world
+`````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('no panic on badly positioned fences', () {
+      final content = chomp('''
+````javascript
+// hello world
+```
+asdf
+````
+```python
+''');
+      final expected = chomp('''
+`````
+````javascript
+// hello world
+```
+asdf
+````
+```python
+`````
+''');
+      final expectedWithInfoString = chomp('''
+`````info
+````javascript
+// hello world
+```
+asdf
+````
+```python
+`````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('code blocks marked by indentation or tilde fences don\'t affect result', () {
+      final content = chomp('''
+    // hello world
+
+~~~~~~
+code
+~~~~~~
+''');
+      final expected = chomp('''
+```
+    // hello world
+
+~~~~~~
```

Comment:
Then I think a further improvement to the test code would be to use a test helper to deduplicate between `content` and `expected`, probably in all of the test cases. That would be a further big reduction in the repetitiveness of the cases, making it easier to scan through to understand what they're all doing.

The key here is that `content` is straightforward to infer from `expected`. So the test helper can take `expected` as an argument; remove the first and last lines; pass that to `wrapWithBacktickFence`; and just check that the result equals the original `expected`.

Then each of these test cases will have just a single copy of the string, and a simple function call around it.